File: fsharp/yacht/Yacht.fs
module Yacht

type Category = 
    | Ones
    | Twos
    | Threes
    | Fours
    | Fives
    | Sixes
    | FullHouse
    | FourOfAKind
    | LittleStraight
    | BigStraight
    | Choice
    | Yacht

type Die =
    | One 
    | Two 
    | Three
    | Four 
    | Five 
    | Six

let dice_sum dice =
    dice |> Seq.map(fun die -> 
        match die with
        | Die.One -> 1
        | Die.Two -> 2
        | Die.Three -> 3
        | Die.Four -> 4
        | Die.Five -> 5
        | Die.Six -> 6
    ) |> Seq.sum

let count dice die number =
    let scored = dice |> Seq.filter (fun x -> x = die) |> Seq.length 
    scored * number

let score_yacht dice =
    let distinct = dice |> Seq.distinct |> Seq.length
    if distinct = 1 then 50 else 0

let score_full_house dice =
    match List.sort(dice) with
    | [a; b; c; d; e] when a = b && a <> c && c = d && d = e -> dice_sum dice
    | [a; b; c; d; e] when a = b && b = c && c <> d && d = e -> dice_sum dice
    | _ -> 0

let score_four_of_a_kind dice =
    match List.sort(dice) with
    | [a; b; c; d; e] when a <> b && b = c && c = d && d = e -> dice_sum [b; c; d; e]
    | [a; b; c; d; e] when a = b && b = c && c = d && d <> e -> dice_sum [a; b; c; d]
    | [a; b; c; d; e] when a = b && b = c && c = d && d = e -> dice_sum [a; b; c; d]
    | _ -> 0

let score_little_straight dice = 
    match List.sort(dice) with
    | [Die.One; Die.Two; Die.Three; Die.Four; Die.Five] -> 30
    | _ -> 0

let score_big_straight dice = 
    match List.sort(dice) with
    | [Die.Two; Die.Three; Die.Four; Die.Five; Die.Six] -> 30
    | _ -> 0


    
let score category dice = 
    match category with
    | Category.Yacht -> score_yacht dice
    | Category.FullHouse -> score_full_house dice
    | Category.FourOfAKind-> score_four_of_a_kind dice
    | Category.LittleStraight -> score_little_straight dice
    | Category.BigStraight -> score_big_straight dice
    | Category.Choice -> dice_sum dice
    | Category.Ones -> count dice Die.One 1 
    | Category.Twos -> count dice Die.Two 2
    | Category.Threes -> count dice Die.Three 3
    | Category.Fours -> count dice Die.Four  4
    | Category.Fives -> count dice Die.Five 5
    | Category.Sixes -> count dice Die.Six 6

File: python/yacht/yacht.py
# Score categories.
# Change the values as you see fit.
YACHT = 0
ONES = 1
TWOS = 2
THREES = 3
FOURS = 4
FIVES = 5
SIXES = 6
FULL_HOUSE = 7
FOUR_OF_A_KIND = 8
LITTLE_STRAIGHT = 9
BIG_STRAIGHT = 10
CHOICE = 11


def score(dice, category):

    dice.sort()

    if category == YACHT:
        return score_yacht(dice)

    if category == CHOICE:
        return sum(dice)
    
    if category == BIG_STRAIGHT:
        return 30 if dice == [2, 3, 4, 5, 6] else 0

    if category == LITTLE_STRAIGHT:
        return 30 if dice == [1, 2, 3, 4, 5] else 0

    if category == FOUR_OF_A_KIND:
        return score_four_of_a_kind(dice)
    
    if category == FULL_HOUSE:
        return score_full_house(dice)
    
    return score_categories(dice, category)

def score_yacht(dice):
    return 50 if len(set(dice)) == 1 else 0

def score_categories(dice, category):
    return category * dice.count(category)

def score_four_of_a_kind(dice):
    elem_0 = dice[0]
    elem_1 = dice[1]
    if dice.count(elem_0) >= 4:
        return elem_0 * 4
    
    if dice.count(elem_1) >= 4:
        return elem_1 * 4

    return 0

def score_full_house(dice):
    elem = dice[0]

    if len(set(dice)) == 2 and (dice.count(elem) == 2 or dice.count(elem) == 3):
        return sum(dice)

    return 0
